pdd/PDD-multistream.md

```diff
@@ -0,0 +1,139 @@
+# PDD for multistream (example/use case)
+
+## Protocol
+
+The motivation and base of the protocol can be found here:
+
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream---self-describing-protocol-stream
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream-selector---self-describing-protocol-stream-selector
+
+The multistream protocol does not cover:
+
+- discovering participants, selecting transports, and establishing connections
+- managing the state of the connection
+
+multistream enables several types of streams to be used over one single stream, like an intelligent message broker that offers the ability to negotiate the protocol and version that is going to be used. To simplify, a visual representation can be:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ─ ─ ┐┌ ─ ┐
+ dht-id/1.0.1│ bitswap/1.2.3  ...
+└ ─ ─ ─ ─ ─ ─ └ ─ ─ ─ ─ ─ ─ ┘└ ─ ┘
+┌────────────────────────────────┐
+│ multistream-select             │
+└────────────────────────────────┘
+┌────────────────────────────────┐
+│transport                       │
+│  [TCP, UDP, uTP, ...]          │
+└────────────────────────────────┘
+```
+
+multistream doesn't cover stream multiplexing over the same connection, however, we can achieve this by leveraging functionality offered by SPDY or HTTP/2.
+
+```
+┌─── ───┐┌─── ───┐┌ ┌─── ───┐┌─── ───┐┌┐
+ a/1.0.0  b/1.0.0  │ a/1.0.0  b/1.0.0 ││
+└─── ───┘└─── ───┘└ └─── ───┘└─── ───┘ ┘
+┌─── ──── ──── ──── ┌─── ──── ──── ──── ┌───
+│multistream-select││multistream-select││...│
+└ ──── ──── ──── ──┘└ ──── ──── ──── ──┘└ ──┘
+┌─── ──── ──── ──── ──── ──── ──── ──── ──── ┌───
+│stream multiplexing                        ││...│
+│ [HTTP/2, SPDY]                            ││   │
+└─ ──── ──── ──── ──── ──── ──── ──── ──── ─┘└─ ─┘
+┌────────────────────────────────────────────────┐
+│ multistream-select                             │
+└────────────────────────────────────────────────┘
+┌────────────────────────────────────────────────┐
+│transport                                       │
+│  [TCP, UDP, uTP, ...]                          │
+└────────────────────────────────────────────────┘
+```
+
+multistream messages have the following structure:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+ ┌───────────────┐┌───────┐│
+││varint with    ││message│
+ │ message length││       ││
+│└───────────────┘└───────┘
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
```

Comment:
On Fri, Jun 12, 2015 at 04:45:29PM -0700, Juan Batiz-Benet wrote:
> one comment: we don't need this frame for _every_ message. the
> multistream idea is just to make "multistream protocols" that all
> begin with the recognizable header…

That's enough for the test-suite's multistream-select and multistream
drivers (that get messages from the transcript and encode them
correctly for those protocols).

> … but after that could pipe your own thing.

So the test suite will also need protocol drivers for any protocols
its testing.  Each per-protocol driver is responsible for testing both
the syntactic and semantic correctness of the testee's implementation
of that protocol, and the interface for a per-protocol driver has to
accept messages that the test suite is reading from the transcript
(e.g. ‘hey, how is it going?’) and return messages that it receives
from the testee.